packages/react-components/src/tree/tree.tsx

```diff
@@ -0,0 +1,189 @@
+import {
+  useCallback,
+  useEffect,
+  useId,
+  useMemo,
+  useState,
+  type FC,
+  type KeyboardEvent,
+} from "react";
+import { useControllableValue } from "../hooks.js";
+import { useTreeControls } from "./tree-control.js";
+import { TreeViewRow } from "./tree-row.js";
+import style from "./tree.module.css";
+import type { TreeNode, TreeRow } from "./types.js";
+
+export interface TreeProps<T extends TreeNode> {
+  /**
+   * If tree allows keeping a current selection.
+   * @default no selection.
+   */
+  readonly selectionMode?: "none" | "single";
+  readonly tree: T;
+  readonly nodeIcon?: FC<{ node: T }>;
+  readonly selected?: string;
+  readonly onSelect?: (id: string) => void;
+  readonly expanded?: Set<string>;
+  readonly onSetExpanded?: (id: Set<string>) => void;
+}
+
+export function Tree<T extends TreeNode>({
+  tree,
+  selected,
+  onSelect,
+  onSetExpanded,
+  nodeIcon,
+  selectionMode = "none",
+}: TreeProps<T>) {
+  const id = useId();
+  const { expanded, toggleExpand, expand, collapse, renderSignal } = useTreeControls({
+    onSetExpanded,
+  });
+  const [focusedIndex, setFocusedIndex] = useState(-1);
+  const [selectedKey, setSelectedKey] = useControllableValue(selected, undefined, onSelect);
+
+  const rows = useMemo(
+    () => getTreeRowsForNode(expanded, toggleExpand, tree),
+    [renderSignal, toggleExpand, tree]
```

Comment:
Not sure why `renderSignal` is included, but anything referenced by the hook should be in the deps array
```suggestion
    [expanded, toggleExpand, tree]
```

Reply:
its actually due to some optimization I have been doing where we reuse teh same `Set` to keep track of expanded node and use this renderSignal as a way to keep track of it but maybe i can do something else

Reply:
ok fixed to just clone the set when the internal signal change, we can do further optimization later if the needs arise